drivers/soundwire/generic_bandwidth_allocation.c

```diff
@@ -43,9 +43,8 @@ static void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
 	int port_bo, sample_int;
 	unsigned int rate, bps, ch = 0;
 
-	port_bo = t_data->block_offset;
-
 	list_for_each_entry(s_rt, &m_rt->slave_rt_list, m_rt_node) {
+		port_bo = t_data->block_offset;
```

Comment:
@RanderWang we should make this conditional.

if you have two slaves involved with the same stream, then on playback you can
a) read the same bitsSlots for all slaves, and extract the relevant information inside the slave (your solution, which works for playback)
b) read different bitSlots for all slaves, and you need to play with the block offset (the existing solution)

For capture, you have no choice. The Slaves *have* to use different slots, so your solution would not work - it would generate bus conflicts whenever one Slave drives a Logic1 while the other tries to drive a Logic0.

In addition we need to make sure the block offset and channel masks are aligned. This may be why you can't get audio on your second speaker. I honestly don't know why the existing solution doesn't work for you.

Reply:
@plbossart  The channel masks are set by codec driver, .e.g. 1308sdw https://github.com/thesofproject/linux/blob/upstream/soundwire/sound/soc/codecs/rt1308-sdw.c#L546.  The masks are 0x3 for all the 1308sdw.  In my PR: https://github.com/thesofproject/linux/pull/1677, I advised using tdm slot to change mask.

For a 2 channels stream with two slaves
slave0 dp offset = t_data->block_offset
slave1 dp offset = t_data->block_offset + 2 * bps.
But now master will send only two channels data in frame, so slave0 will get data, but slave1 will get zero data.

Reply:
> slave0 dp offset = t_data->block_offset
> slave1 dp offset = t_data->block_offset + 2 * bps.

ok, thanks for the explanations, I am starting to understand the problem.
What happens if you use (even with a hack?)

slave0 dp offset = t_data->block_offset
slave1 dp offset = t_data->block_offset + 1 * bps.

and maybe hardcode 1ch for each port as well?

Reply:
@plbossart  That is what https://github.com/thesofproject/linux/pull/1677 did.  use tdm slot to force each slave using one channel. playback works with this solution.  For 1308, we also need to set a mixer controller to force use channel 0 for each 1308, since 1308 only supports 2 channels. In this case, one 1308 uses channel 0 in master and discard channel1, one 1308 uses channel1 in master, discard the zero data after channel1

Reply:
Now that I think of it, the set_tdm_slot() function is actually a good match for SoundWire. We have conveniently 8 channels supported, so can set the mask as needed. The only tricky part is that the ports are not full-duplex, so we'd need to have zeroes in the TX slots when the port is a sink - but that's no big deal.

so if we use the set_tdm_slot() function, we can support this mode as PR #1677 suggests it. It also works for capture.

But we can't do what I suggested if we want both slaves to fetch the same data, e.g. if we set the slot 0/1 for both slaves, the offset value won't work for Slave1 - unless we also modify the bandwidth allocation to take the ch_mask into account and reset tbe block offset.

Gah, this is a nice problem. Need more coffee.